src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -321,107 +300,107 @@ private bool TryLookupLengthOrCount(SyntaxNode node, TypeSymbol inputType, out P
-                if (boundAccess is BoundIndexerAccess { ResultKind: LookupResultKind.Viable } indexerAccess &&
-                    indexerAccess.Indexer.GetMethod is { } getMethod && IsAccessible(getMethod, ref useSiteInfo))
-                {
-                    // PROTOTYPE(list-patterns) Can this be ever true? If so, move to if above
-                    Debug.Assert(!indexerAccess.Indexer.IsStatic);
-                    result = BindIndexerDefaultArguments(indexerAccess, BindValueKind.RValue, bindingDiagnostics);
-                    _ = GetWellKnownTypeMember(argIsIndex ? WellKnownMember.System_Index__ctor : WellKnownMember.System_Range__ctor, bindingDiagnostics, syntax: syntax);
-                    // At this point we have succeeded to bind a viable indexer,
-                    // report additional binding diagnostics that we have seen so far
-                    diagnostics.AddRange(bindingDiagnostics);
+                    var analyzedArguments = AnalyzedArguments.GetInstance();
+                    analyzedArguments.Arguments.Add(new BoundIndexOrRangeIndexerPatternValuePlaceholder(syntax, argType));
+                    var receiver = new BoundImplicitReceiver(syntax, receiverType);
+                    BoundExpression boundAccess = BindIndexerOrIndexedPropertyAccess(syntax, receiver, indexerGroup, analyzedArguments, bindingDiagnostics);
+                    switch (boundAccess)
+                    {
+                        case BoundIndexerAccess { ResultKind: LookupResultKind.Viable, Indexer: { GetMethod: { } getMethod } } boundIndexerAccess when IsAccessible(getMethod, ref useSiteInfo):
+                            lookupResult.Clear();
+                            if (!TryLookupLengthOrCount(receiverType, lookupResult, out lengthProperty, ref useSiteInfo))
+                                goto default;
+                            ReportDiagnosticsIfObsolete(diagnostics, lengthProperty, syntax, hasBaseReceiver: false);
+                            // PROTOTYPE(list-patterns) Can this be ever true? If so, move to if above
+                            Debug.Assert(!boundIndexerAccess.Indexer.IsStatic);
+                            indexerAccess = BindIndexerDefaultArguments(boundIndexerAccess, BindValueKind.RValue, bindingDiagnostics);
+                            _ = GetWellKnownTypeMember(argIsIndex ? WellKnownMember.System_Index__ctor : WellKnownMember.System_Range__ctor, bindingDiagnostics, syntax: syntax);
+                            found = true;
+                            break;
+                        case BoundIndexOrRangePatternIndexerAccess boundIndexOrRangePatternIndexerAccess:
+                            patternSymbol = boundIndexOrRangePatternIndexerAccess.PatternSymbol;
+                            lengthProperty = boundIndexOrRangePatternIndexerAccess.LengthOrCountProperty;
+                            found = true;
+                            break;
+                        default:
```

Comment:
As far as I can tell, we only reach this case from the `goto default` above. Is that correct? If so, let's remove the `goto` and throw instead here.